### PR TITLE
Introducing product type constants

### PR DESCRIPTION
**Pull request by wjrosa**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces constants for all product types. The idea is to make it easier to reference them, avoid typos, make the code stricter, make it easier to find status usages, centralize them, improve documentation, and many other reasons that should help developers create code related to them.

### How to test the changes in this Pull Request:

- Checkout and build this branch on your test environment (`dev/introducing-product-type-constants`)
- Perform some smoke tests such as:
    - Store setup
    - Payment gateway installation (Stripe, for example)
    - Product creation
    - Common checkout
    - Shortcode checkout
    - My account page navigation
    - Order listing/details
- Confirm that no regression was introduced

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Introduces and implements product type constants in the WooCommerce plugin codebase.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/admin/class-wc-admin-post-types.php [hunk 1]
```diff
@@ -577,8 +578,14 @@ public function bulk_edit_save( $post_id, $product ) {
...
-		// Handle price - remove dates and set to lowest.
-		$change_price_product_types    = apply_filters( 'woocommerce_bulk_edit_save_price_product_types', array( 'simple', 'external' ) );
+		/**
+		 * Handle price - remove dates and set to lowest.
+		 *
+		 * @param array $product_types Array of product types that can change price.
+		 *
+		 * @since 3.0.0
+		 */
+		$change_price_product_types    = apply_filters( 'woocommerce_bulk_edit_save_price_product_types', array( ProductType::SIMPLE, ProductType::EXTERNAL ) );
```
Comment (barryhughes) [inline plugins/woocommerce/includes/admin/class-wc-admin-post-types.php:588]: Thanks for adding the docblocks!

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php [hunk 2]
```diff
@@ -489,8 +490,8 @@ protected function get_product_data( $product ) {
 			'download_limit'        => $product->get_download_limit(),
 			'download_expiry'       => $product->get_download_expiry(),
 			'download_type'         => 'standard',
-			'external_url'          => $product->is_type( 'external' ) ? $product->get_product_url() : '',
-			'button_text'           => $product->is_type( 'external' ) ? $product->get_button_text() : '',
+			'external_url'       => $product->is_type( ProductType::EXTERNAL ) ? $product->get_product_url() : '',
+			'button_text'        => $product->is_type( ProductType::EXTERNAL ) ? $product->get_button_text() : '',
```
Comment (barryhughes) [inline plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php:494]: Something's funky about the whitespace before `=>` here.

File: plugins/woocommerce/templates/single-product/meta.php [hunk 3]
```diff
@@ -12,9 +12,11 @@
  *
  * @see         https://woocommerce.com/document/template-structure/
  * @package     WooCommerce\Templates
- * @version     3.0.0
+ * @version     3.0.1
```
Comment (barryhughes) [inline plugins/woocommerce/templates/single-product/meta.php]: Here and elsewhere, is the 'minor' bump a means of satisfying the template version bump check (but without actually bumping it to `9.7.0`)?

**Review by barryhughes — COMMENTED** [review]
Thanks, looking good! I left some minor comments/questions. Also wanted to check that we feel like we've caught everything.

```sh
grep -Ern "['\"](external|grouped|simple|variable|variation)['\"]" ./ | grep .php
```

Using the above one-liner surfaces quite a few lines we might want to check on. I would bet a lot of them can be ignored, but some [(example)](https://github.com/woocommerce/woocommerce/blob/66a908abc35fb64c1d496838e95ed5c48256c46c/plugins/woocommerce/src/Blocks/BlockTypes/ProductStockIndicator.php#L47) *look* like they could be valid matches.

File: plugins/woocommerce/templates/single-product/meta.php (hunk 3 above)
Comment (wjrosa) [reply plugins/woocommerce/templates/single-product/meta.php]: Oops, I thought the version referred to the file itself, not the plugin version. I will update all occurrences accordingly 👍

Comment (wjrosa) [reply plugins/woocommerce/templates/single-product/meta.php]: Both updated ✅

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php (hunk 2 above)
Comment (wjrosa) [reply plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php:494]: Yeah. That was the only way linting passed 🤔 . Not sure how to proceed here.

**Comment by wjrosa** [comment]
> Thanks, looking good! I left some minor comments/questions. Also wanted to check that we feel like we've caught everything.
> 
> ```shell
> grep -Ern "['\"](external|grouped|simple|variable|variation)['\"]" ./ | grep .php
> ```
> 
> Using the above one-liner surfaces quite a few lines we might want to check on. I would bet a lot of them can be ignored, but some [(example)](https://github.com/woocommerce/woocommerce/blob/66a908abc35fb64c1d496838e95ed5c48256c46c/plugins/woocommerce/src/Blocks/BlockTypes/ProductStockIndicator.php#L47) _look_ like they could be valid matches.

Thanks for reviewing it, Barry! I replaced this occurrence and some others in https://github.com/woocommerce/woocommerce/pull/53759/commits/d1bc95ac7637ed73f70e80474122feda50670020

File: plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php (hunk 2 above)
Comment (barryhughes) [reply plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php:494]: Hmm. Yeah, running `phpcs` on the whole file reveals lots of problems with spacing and the `WordPress.Arrays.MultipleStatementAlignment.DoubleArrowNotAligned` rule in particular, so perhaps the whole file needs a bit of re-formatting. I don't think we should do that in this PR, though. Let's ignore for the time being.

**Review by barryhughes — APPROVED** [review]
Nice, nothing else to add from my end: thanks for taking care of this!

Approving but not merging (in case we want someone else to also take a look ... that's not essential, but on the other hand it is a pretty big change) :+1:

**Pull request merged**